src/config/env.py

```diff
@@ -0,0 +1,105 @@
+"""
+env.py
+
+This module is responsible for loading, parsing, and validating the application's
+environment configuration from the `env.yaml` file. It uses Pydantic to ensure that
+the configuration adheres to expected formats and types.
+
+Functions:
+- create_default_config: Creates a default configuration file if it doesn't exist.
+
+Classes:
+- EnvConfig: Loads the configuration and exposes it as Python objects.
+- RemoteNode: Represents a remote node configuration with validation.
+- Config: Represents the overall configuration structure with validation.
+
+Usage:
+- Initialize an instance of EnvConfig to load the configuration.
+- Access configuration values via properties (e.g., env.remote_nodes).
+"""
+
+from pathlib import Path
+from typing import Dict, List
+
+import yaml
+from pydantic import BaseModel, HttpUrl, ValidationError
+
+ENV_PATH = Path(__file__).resolve().parent.parent.parent / "env.yaml"
+
+
+def create_default_config():
+    """Creates a default configuration file if it doesn't exist."""
+    default_config = """
+# This file contains environment-specific configurations for the application.
+# It is git-ignored to keep local secrets safe.
+
+# The configuration stores Ethereum RPC node details.
+remote_nodes:
+  - name: mainnet_archive
+    # Replace with your Ethereum RPC node URL
+    node_url: http://example.com
+    # Optional: Headers for RPC requests
+    rpc_headers:
+      client-secret: <secret>
+"""
+    with ENV_PATH.open("w") as file:
+        file.write(default_config.strip())
+        print(f"Env file created: {ENV_PATH}")
+
+
+class RemoteNode(BaseModel):
+    """
+    Represents a configuration for a remote node.
+
+    Attributes:
+    - name (str): The name of the remote node.
+    - node_url (HttpUrl): The URL for the remote node, validated as a proper URL.
+    - rpc_headers (Dict[str, str]): A dictionary of optional RPC headers, defaults to empty dict.
+    """
+
+    name: str
+    node_url: HttpUrl
+    rpc_headers: Dict[str, str] = {}
+
+
+class Config(BaseModel):
```

Comment:
Just FYI: We could consider using pydantic's `BaseSettings` class in the future.
https://docs.pydantic.dev/latest/concepts/pydantic_settings/#settings-management

I'm not quite sure how pydantic would work with the yaml though? Pydantic is cable of creating a schema, which could be loaded in VS code to help users edit config files. I'm not sure this would work with yaml, but it does look like we could use pydantic's `YamlConfigSettingsSource`. A schema is overkill for now though imo :-)

Reply:
I reviewed it I believe for our usecase this has diminishing returns. `pydantic-settings` will try merge environment the current runtime or an .env file.  We don't need either either. This also takes performance hit and the possibility of collusion should be there be an environment variable with the same name available in a user's runtime.

https://github.com/pydantic/pydantic-settings/blob/a72fa7383f4436148db01595b2993b846c91d3fc/pydantic_settings/main.py#L326-L333